src/Storages/StorageAggregatingMemory.cpp

```diff
@@ -0,0 +1,513 @@
+    }
+};
+
+StorageAggregatingMemory::StorageAggregatingMemory(
+    const StorageID & table_id_,
+    ConstraintsDescription constraints_,
+    const ASTCreateQuery & query,
+    ContextPtr context_)
+    : IStorage(table_id_),
+      log(&Poco::Logger::get("StorageAggregatingMemory"))
+{
+    if (!query.select)
+        throw Exception("SELECT query is not specified for " + getName(), ErrorCodes::INCORRECT_QUERY);
+
+    if (query.select->list_of_selects->children.size() != 1)
+        throw Exception("UNION is not supported for AggregatingMemory", ErrorCodes::INCORRECT_QUERY);
+
+    // TODO check validity of aggregation query inside this func
+    select_query = SelectQueryDescription::getSelectQueryFromASTForAggr(query.select->clone());
+    constructor_context = Context::createCopy(context_);
+    constructor_constraints = constraints_;
+}
+
+void StorageAggregatingMemory::lazyInit()
+{
+    if (is_initialized)
+        return;
+
+    std::lock_guard lock(mutex);
+
+    if (is_initialized)
+        return;
+
+    ASTPtr select_ptr = select_query.inner_query;
+
+    auto select_context = constructor_context;
+
+    /// Get info about source table.
+    JoinedTables joined_tables(constructor_context, select_ptr->as<ASTSelectQuery &>());
+    source_storage = joined_tables.getLeftTableStorage();
+    NamesAndTypesList source_columns = source_storage->getInMemoryMetadata().getColumns().getAll();
+
+    ColumnsDescription columns_before_aggr;
+    for (const auto & column : source_columns)
+    {
+        ColumnDescription column_description(column.name, column.type);
+        columns_before_aggr.add(column_description);
+    }
+
+    /// Get list of columns we get from select query.
+    Block header = InterpreterSelectQuery(select_ptr, select_context, SelectQueryOptions().analyze()).getSampleBlock();
+
+    ColumnsDescription columns_after_aggr;
+
+    /// Insert only columns returned by select.
+    for (const auto & column : header)
+    {
+        ColumnDescription column_description(column.name, column.type);
+        columns_after_aggr.add(column_description);
+    }
+
+    StorageInMemoryMetadata storage_metadata;
+    storage_metadata.setColumns(std::move(columns_after_aggr));
+    storage_metadata.setConstraints(constructor_constraints);
+    storage_metadata.setSelectQuery(select_query);
+    setInMemoryMetadata(storage_metadata);
+
+    StorageInMemoryMetadata src_metadata;
+    src_metadata.setColumns(std::move(columns_before_aggr));
+    src_block_header = src_metadata.getSampleBlock();
+
+    src_metadata_snapshot = std::make_shared<StorageInMemoryMetadata>(src_metadata);
+
+    Names required_result_column_names;
+
+    auto syntax_analyzer_result
+        = TreeRewriter(select_context)
+              .analyzeSelect(
+                  select_ptr, TreeRewriterResult(src_block_header.getNamesAndTypesList()), {}, {}, required_result_column_names, {});
+
+    query_analyzer = std::make_unique<SelectQueryExpressionAnalyzer>(
```

Comment:
`query_analyzer` and `analysis_result` are not used anywhere else except this function.
So, we can remove it from StorageAggregatingMemory members